### DISCUSSION TITLE
GetAll Products

### DISCUSSION THREAD

**Discussion by szakalo** [XiaoFaye/WooCommerce.NET#728] [Q&A]
Hi
How to download a list of all products. When I use

`var products = await wc.Product.GetAll();`

only gets a list of 10 products. I have over 1000 of them in my WooCommerce.

**Comment by tiziana-git** [comment]
same problem here

**Comment by omergokdere** [comment]
Create an object of Product and convert it into List with Task you can get em all.

here is the sample 

public async Task<List<Product>> getAllProductsV2()
        {
            Product product = new();
            List<Product> allProducts = new List<Product>();
            try
            {
                allProducts = await wc.Product.GetAll();
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine("ERROR MESSAGE: " + ex.Message);
                System.Diagnostics.Debug.WriteLine("ERROR STACK TRACE: " + ex.StackTrace);
            }
            return allProducts;
        }

**Reply by nikolamarkovic1987** [reply]
This approach returns only 10 products....

**Comment by danieledignani** [comment]
> Hi How to download a list of all products. When I use
> 
> `var products = await wc.Product.GetAll();`
> 
> only gets a list of 10 products. I have over 1000 of them in my WooCommerce.

I had the same issue. I was going deeper and search for a solution. With postman I understand that the get request
https://yoursite/wp-json/wc/v3/products
return only 10 products as in your case. You could add the parameter per_page for a maximum of 100, and another parameter page to iterate each page. That's a woocommerce limitation and not a plugin issue.
So the request should be 
https://yoursite/wp-json/)wc/v3/products?per_page=100&page=X
where X is the page number
We need to do that in c# code and that's what I did to solve my problem. I create this function
```
// You could change this constant if the server response is too slow
private const int WOOCOMMERCE_ITEMS_PER_PAGE = 100;

private async Task<List<Product>> GetAllProducts()
{
    List<Product> products = new List<Product>();
    Dictionary<string, string> filters = new Dictionary<string, string>();

    var pageToRetrive = 0;
    filters.Add("post_status", $"any");
    filters.Add("per_page", $"{WOOCOMMERCE_ITEMS_PER_PAGE}");
    filters.Add("page", $"{pageToRetrive}");

    int lastProductsRetrivedCount;
    do
    {
        pageToRetrive++;
        filters["page"] = $"{pageToRetrive}";
        var productsInPage = await wc.Product.GetAll(filters);
        products.AddRange(productsInPage);
        lastProductsRetrivedCount = productsInPage.Count;
    } while (lastProductsRetrivedCount == WOOCOMMERCE_ITEMS_PER_PAGE);

    // This code below could be not necessary for you
    var variableProducts = products.Where(p => p.type == "variable");
    if (variableProducts.Count() == 0)
    {
        return products;
    }
    var variableProductsIds = variableProducts.SelectMany(p => p.variations);
    var productVariablesToAdd = new List<Product>();
    foreach (var variableId in variableProductsIds)
    {
        var variableProduct = await wc.Product.Get((ulong)variableId);
        if (variableProduct != null)
        {
            productVariablesToAdd.Add(variableProduct);
        }
    }
    products.AddRange(productVariablesToAdd);
    return products;
}
```
